src/codegen/mod.rs

```diff
@@ -75,20 +80,69 @@ pub trait CodeGen<Tgt: Target> {
             println!("{}", String::from_utf8_lossy(&clang_proc.stderr));
         }
 
-        Ok(BuiltArtifact::new(binary_path, source_path, dirname, None))
+        Ok(BuiltArtifact::new(
+            binary_path,
+            source_path,
+            dirname,
+            bench_samples,
+        ))
+    }
+
+    /// Benchmark several times, returning the minimum of inner loop means.
+    //
+    //  This will first estimate a good number of inner loop iterations, then
+    //  build an executable which loops that number of times, returning the mean.
+    //  The final `result` computed is the minimum of the means after running
+    //  that executable `repeat` times.
+    fn time_impl_robustly(
+        &self,
+        print_code: bool,
+        repeat: Option<usize>,
+    ) -> Result<RobustTimingResult> {
+        let repeat = repeat.unwrap_or(10); // default: 10
+
+        // Collect a single rough sample.
+        let time_check_artifact = self.build(false, Some(1))?;
+        let rough_secs = time_check_artifact.measure_time()?;
+
+        // Choose a good number of iterations for benchmarks' inner loop.
+        let inner_iters = max(
+            MIN_SAMPLES,
+            (MIN_TRIAL_TIME_SECS / rough_secs).ceil() as usize,
+        );
+        info!("Goal iterations: {inner_iters}");
+
+        // Run main benchmark loop.
+        let artifact = self.build(print_code, Some(inner_iters))?;
+        let mut means = Vec::with_capacity(repeat);
+        for _ in 0..repeat {
+            let secs = artifact.measure_time()?;
+            info!("Sample runtime result {secs}s");
+            means.push(secs);
+        }
+
+        Ok(RobustTimingResult {
+            result: *means
+                .iter()
+                .min_by(|a, b| a.partial_cmp(b).unwrap())
```

Comment:
This will fail if floats are `inf` or `nan`. Rather than checking for that case, have `measure_time` return a `Duration`, which is `Eq`.